multicluster/cmd/gateways.go

```diff
@@ -1,95 +1,218 @@
+	io_prometheus_client "github.com/prometheus/client_model/go"
+	"github.com/prometheus/common/expfmt"
 	"github.com/spf13/cobra"
 	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
 type (
 	gatewaysOptions struct {
-		gatewayNamespace string
+		clusterName string
+		wait        time.Duration
+	}
+
+	gatewayMetrics struct {
+		clusterName string
+		metrics     []byte
+		err         error
+	}
+
+	gatewayStatus struct {
 		clusterName      string
-		timeWindow       string
+		alive            bool
+		numberOfServices int
+		latency          uint64
 	}
 )
 
+func newGatewaysOptions() *gatewaysOptions {
+	return &gatewaysOptions{
+		wait: 30 * time.Second,
+	}
+}
+
 func newGatewaysCommand() *cobra.Command {
 
-	opts := gatewaysOptions{}
+	opts := newGatewaysOptions()
 
 	cmd := &cobra.Command{
 		Use:   "gateways",
 		Short: "Display stats information about the gateways in target clusters",
 		Args:  cobra.NoArgs,
 		RunE: func(cmd *cobra.Command, args []string) error {
-			req := &pb.GatewaysRequest{
-				RemoteClusterName: opts.clusterName,
-				GatewayNamespace:  opts.gatewayNamespace,
-				TimeWindow:        opts.timeWindow,
-			}
-
 			k8sAPI, err := k8s.NewAPI(kubeconfigPath, kubeContext, impersonate, impersonateGroup, 0)
 			if err != nil {
 				return err
 			}
 
-			ctx := cmd.Context()
-
-			vizNs, err := k8sAPI.GetNamespaceWithExtensionLabel(ctx, vizCmd.ExtensionName)
+			// Get all the service mirror components in the
+			// linkerd-multicluster namespace which we'll collect gateway
+			// metrics from.
+			multiclusterNs, err := k8sAPI.GetNamespaceWithExtensionLabel(cmd.Context(), MulticlusterExtensionName)
 			if err != nil {
-				return fmt.Errorf("make sure the linkerd-viz extension is installed, using 'linkerd viz install' (%w)", err)
+				return fmt.Errorf("make sure the linkerd-multicluster extension is installed, using 'linkerd multicluster install' (%w)", err)
 			}
-
-			client, err := client.NewExternalClient(ctx, vizNs.Name, k8sAPI)
+			selector := fmt.Sprintf("component=%s", "linkerd-service-mirror")
+			pods, err := k8sAPI.CoreV1().Pods(multiclusterNs.Name).List(cmd.Context(), metav1.ListOptions{LabelSelector: selector})
 			if err != nil {
-				return err
-			}
-
-			resp, err := requestGatewaysFromAPI(client, req)
-			if err != nil {
-				fmt.Fprintln(os.Stderr, err)
+				fmt.Fprintf(os.Stderr, "failed to list pods in namespace %s: %s", multiclusterNs.Name, err)
 				os.Exit(1)
 			}
 
-			renderGateways(resp.GetOk().GatewaysTable.Rows, stdout)
+			var statuses []gatewayStatus
+			gatewayMetrics := getGatewayMetrics(k8sAPI, pods.Items, opts.wait)
+			for _, gateway := range gatewayMetrics {
+				if gateway.err != nil {
+					fmt.Fprintf(os.Stderr, "Failed to get gateway status for %s: %s\n", gateway.clusterName, gateway.err)
+					continue
+				}
+				if opts.clusterName != "" && gateway.clusterName != opts.clusterName {
+					continue
+				}
```

Comment:
This is a bit nitpicky so take it or leave, but we could perhaps do this check before we retrieve the service-mirrors? 

Each service mirror should have the clusterName as part of its labels (`mirror.linkerd.io/cluster-name: {{.Values.targetClusterName}}`). When a cluster name is provided in the options, we can add it to the label selector. This would allow us to only process metrics for the cluster we want and possibly be a bit more efficient?

e.g

```go
selector := fmt.Sprintf("component=%s", "linkerd-service-mirror")
if opts.clusterName != "" {
    selector := fmt.Sprintf("%s,mirror.linkerd.io/cluster-name=%s", selector, opts.clusterName)
}
...

And then we won't have to deal with the condition in the loop. Wdyt?

Reply:
Yep that sounds good 👍